### PR TITLE
Cache default visualization responses for all studies (SCP-2921, SCP-3308)

### PR DESCRIPTION
**Pull request by bistline**
As part of the effort to improve performance and reduce overall visualization times, this update changes the behavior for generating/warming visualization caches.  Previously, a response would only be cached after first having been requested by a user, and that cache would live only until the next deployment, or until the study owner updated a file or setting that would cause the cache to be cleared.  Now, upon successful ingest of a cluster/metadata file, all "default" visualization caches will be stored proactively, and all visualization caches will persist indefinitely (i.e. through deployments).  If a user updates a default setting that would cause the default cache to be cleared, it is then immediately regenerated.  Furthermore, the defaults will be cached at both the default subsample and "All Cells". 

Note: this _only_ applies to the default cluster & annotation for each study.  In addition, this update refactors the logic for generating visualization cache file paths into the `RequestUtils` class.

Follow-on work to this update will be to measure the improvement to TTI (time to interactive) on the explore tab in the study overview page, and a determination as to whether we want to extend this behavior to all non-expression visualization caches.

This PR satisfies SCP-2921 and portions of SCP-3308.

### CODE REVIEW

**Review by eweitz — APPROVED** [review]
This cache warming is a huge optimization!  I suspect it will drop `perfTime:backend` by > 95% for most scatter plot API requests.

Below, I suggest a few non-blocking observability and readability improvements.

Regarding time-to-interactive analytics: as a separate task, it might help to have a property (say, `nth-log-in-page-view`) that tracks how many times the currently-logged API path has been called in the current page view.  Then you could plot e.g. TTI for Explore = median `timeSincePageLoad` for `scatter:plot` filtered by `nth-call-in-page-view == 1`.  This could be a corollary to adding unique IDs for each page view and search (SCP-3318, which addresses similar but subtly different analytics use cases).

Alternatively, there might be a way to ferret that data out using Mixpanel JQL, which I'll discuss tomorrow in the context of #1028.

File: app/lib/cluster_cache_service.rb [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        annotation_name: CGI.escape(annotation_name), annotation_scope: annotation_scope, annotation_type: annotation_type,
+        subsample: 'all', cluster_name: sanitized_cluster_name
+      }
+      [default_path, named_path].each do |viz_request_path|
+        # specify '_default' for cluster_name to form correct cache path, otherwise use all parameters
+        url_params = viz_request_path.end_with?('_default') ? {cluster_name: '_default'} : full_params
+        cache_path = RequestUtils.get_cache_path(viz_request_path, url_params.with_indifferent_access)
+        viz_data = Api::V1::Visualization::ClustersController.get_cluster_viz_data(study, cluster, url_params)
+        Rails.logger.info "Pre-caching viz data for #{cache_path}"
+        Rails.cache.write(cache_path, viz_data.to_json)
```
[the remark this answers is not in the document]
Comment (bistline) [reply app/lib/cluster_cache_service.rb:49]: In some cases that is true, but for any default cluster that has > 10K points, the stock "default" response will be subsampled, but the other response that gets cached is at the "all cells" level.  That was why I needed to do it inside the block.

File: app/lib/request_utils.rb [hunk 2]
```diff
@@ -1,11 +1,57 @@
-require 'rails/commands/server/server_command'
-
+# RequestUtils: helper class for dealing with request parameters, sanitizing input, and setting
+# cache paths on visualization requests
 class RequestUtils
 
+  # list of parameters to reject from :get_cache_key as they will be represented by request.path
+  # format is always :json and therefore unnecessary
+  CACHE_PATH_BLACKLIST = %w(controller action format study_id)
```
[the remark this answers is not in the document]
Comment (bistline) [reply app/lib/request_utils.rb:7]: Good call - there's actually an industry-wide push for this.  I'll make that update.

File: app/lib/cluster_cache_service.rb [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+      annotation_name, annotation_type, annotation_scope = annotation.split('--')
+      default_path = format_request_path(:api_v1_study_cluster_path, study.accession, '_default')
+      # necessary for legacy cluster names that could contain slashes and other non URL-safe characters
+      sanitized_cluster_name = CGI.escape(cluster.name)
+      named_path = format_request_path(:api_v1_study_cluster_path,study.accession, sanitized_cluster_name)
+      full_params = {
+        annotation_name: CGI.escape(annotation_name), annotation_scope: annotation_scope, annotation_type: annotation_type,
+        subsample: 'all', cluster_name: sanitized_cluster_name
+      }
+      [default_path, named_path].each do |viz_request_path|
```
[the remark this answers is not in the document]
Comment (bistline) [reply app/lib/cluster_cache_service.rb:43]: Good point!

File: app/lib/request_utils.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (bistline) [reply app/lib/request_utils.rb:7]: I'm going to open a ticket for us to do this in the rest of the codebase as well - there's a bunch of places that use terms like this.

[the remark this answers is not in the document]
Comment (bistline) [reply app/lib/request_utils.rb]: 46acbf7

File: app/lib/cluster_cache_service.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (bistline) [reply app/lib/cluster_cache_service.rb]: db6ac6b

File: app/lib/cluster_cache_service.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (devonbush) [reply app/lib/cluster_cache_service.rb:47]: ah, I didn't pick that up on my first read through.  Now I understand what you meant by 'full_params'.  Once we see how this performs in prod, will be interesting to see if this means we can set the default subsample on the explore tab to 'all'

**Review by devonbush — APPROVED** [review]

**Pull request merged**